src/ast/passes/codegen_llvm.cpp

```diff
@@ -151,16 +151,98 @@
+                                               "merge_block",
+                                               parent);
+
+  Value *stack_trace = b_.CreateGetStackScratchMap(stack_type,
+                                                   stack_scratch_failure,
+                                                   loc);
+  b_.CREATE_MEMSET(
+      stack_trace, b_.getInt8(0), uint64_size * stack_type.limit, 1);
+
+  BasicBlock *get_stack_success = BasicBlock::Create(module_->getContext(),
+                                                     "get_stack_success",
+                                                     parent);
+  BasicBlock *get_stack_fail = BasicBlock::Create(module_->getContext(),
+                                                  "get_stack_fail",
+                                                  parent);
+
+  Value *ptr_to_stack_trace = b_.CreatePointerCast(stack_trace,
+                                                   b_.GET_PTR_TY());
+  Value *stack_size = b_.CreateGetStack(
+      ctx_, is_ustack, ptr_to_stack_trace, stack_type, loc);
+  Value *condition = b_.CreateICmpSGE(stack_size, b_.getInt32(uint64_size));
+  b_.CreateCondBr(condition, get_stack_success, get_stack_fail);
+
+  b_.SetInsertPoint(get_stack_fail);
+  if (bpftrace_.debug_output_)
+    b_.CreateDebugOutput("Failed to get stack. Error: %d",
+                         std::vector<Value *>{ stack_size },
+                         loc);
+  b_.CreateStore(b_.getInt64(0), stackid);
+  b_.CreateBr(merge_block);
+  b_.SetInsertPoint(get_stack_success);
+
+  Value *stack_len = b_.CreateUDiv(stack_size, b_.getInt32(uint64_size));
+  // uint64 *stack, uint64 len, uint64 seed
+  AllocaInst *seed = b_.CreateAllocaBPF(CreateUInt64(), "seed");
+  // use the pid as the seed if we're getting user stacks
+  if (is_ustack) {
+    b_.CreateStore(b_.CreateGetPidTgid(loc), seed);
+  } else {
+    b_.CreateStore(b_.getInt64(1), seed);
+  }
+
+  // LLVM-12 produces code that fails the BPF verifier because it
+  // can't determine the bounds of stack_len. The only thing that seems
+  // to work is truncating the type, which is fine because 255 is long enough.
+  Value *trunc_stack_len = b_.CreateTrunc(stack_len, b_.getInt8Ty());
+
+  // Here we use the murmur2 hash function to create the stack ids because
+  // bpf_get_stackid() is kind of broken by design and can suffer from hash
+  // collisions.
+  // More details here: https://github.com/bpftrace/bpftrace/issues/2962
+  Value *murmur_hash_2 = b_.CreateCall(murmur_hash_2_func_,
+                                       { ptr_to_stack_trace,
+                                         trunc_stack_len,
+                                         b_.CreateLoad(b_.getInt64Ty(), seed) },
+                                       "murmur_hash_2");
+  b_.CreateDebugOutput("Stack id %llu. Len: %llu",
+                       std::vector<Value *>{ murmur_hash_2, stack_len },
+                       loc);
```

Comment:
Need a `bpftrace_.debug_output_` guard?

Reply:
oops, good catch.

Reply:
Looks like irbuilderbpf holds onto a reference to bpftrace. Maybe we should just push that guard down into CreateDebugOutput() ?

Reply:
Good idea. Will update